filterExprParser.go

```diff
@@ -0,0 +1,691 @@
+	return !feco.IsNull()
+}
+
+func (feco *FECheckOp) String() string {
+	if feco.IsExists() {
+		return "EXISTS"
+	} else if feco.IsMissing() {
+		return "IS MISSING"
+	} else if feco.IsNotMissing() {
+		return "IS NOT MISSING"
+	} else if feco.IsNull() {
+		return "IS NULL"
+	} else if feco.IsNotNull() {
+		return "IS NOT NULL"
+	} else {
+		return "?? (FECheckOp)"
+	}
+}
+
+// Technically we could have an slice of arguments, but having OneArg vs NoArg vs TwoArg could
+// allow us to do more strict function check (i.e. certain funcs should only allow one argument, etc, at this level)
+type FEConstFuncExpression struct {
+	ConstFuncNoArg   *FEConstFuncNoArg   `@@ |`
+	ConstFuncOneArg  *FEConstFuncOneArg  `@@ |`
+	ConstFuncTwoArgs *FEConstFuncTwoArgs `@@`
+}
+
+func (f *FEConstFuncExpression) String() string {
+	if f.ConstFuncNoArg != nil {
+		return f.ConstFuncNoArg.String()
+	} else if f.ConstFuncOneArg != nil {
+		return f.ConstFuncOneArg.String()
+	} else if f.ConstFuncTwoArgs != nil {
+		return f.ConstFuncTwoArgs.String()
+	} else {
+		return "?? (FEConstFuncExpression)"
+	}
+}
+
+type FEConstFuncNoArg struct {
+	ConstFuncNoArgName *FEConstFuncNoArgName `( @@ "(" ")" )`
+}
+
+func (f *FEConstFuncNoArg) String() string {
+	if f.ConstFuncNoArgName != nil {
+		return fmt.Sprintf("%v()", f.ConstFuncNoArgName.String())
+	} else {
+		return "?? (FEConstFuncNoArg)"
+	}
+}
+
+type FEConstFuncNoArgName struct {
+	Pi *bool `@"PI" |` // FuncPi
+	E  *bool `@"E"`    // FuncE
+}
+
+func (n *FEConstFuncNoArgName) String() string {
+	if n.E != nil && *n.E == true {
+		return "E"
+	} else if n.Pi != nil && *n.Pi == true {
+		return "PI"
+	} else {
+		return "?? (FEConstFuncNoArgName)"
+	}
+}
+
+// Order matters
+type FEConstFuncArgument struct {
+	SubFunc  *FEConstFuncExpression `@@ |`
+	Field    *FEField               `@@ |`
+	Argument *FEValue               `@@`
+}
+
+func (arg *FEConstFuncArgument) String() string {
+	if arg.Argument != nil {
+		return arg.Argument.String()
+	} else if arg.SubFunc != nil {
+		return arg.SubFunc.String()
+	} else if arg.Field != nil {
+		return arg.Field.String()
+	} else {
+		return "?? (FEConstFuncArgument)"
+	}
+}
+
+// Prioritize value over field
+type FEConstFuncArgumentRHS struct {
+	SubFunc  *FEConstFuncExpression `@@ |`
+	Argument *FEValue               `@@ |`
+	Field    *FEField               `@@`
+}
+
+func (arg *FEConstFuncArgumentRHS) String() string {
+	if arg.Argument != nil {
+		return arg.Argument.String()
+	} else if arg.SubFunc != nil {
+		return arg.SubFunc.String()
+	} else if arg.Field != nil {
+		return arg.Field.String()
+	} else {
+		return "?? (FEConstFuncArgument)"
+	}
+}
+
+type FEConstFuncOneArg struct {
+	ConstFuncOneArgName *FEConstFuncOneArgName `( @@ "("`
+	Argument            *FEConstFuncArgument   `@@ ")" )`
+}
+
+func (oa *FEConstFuncOneArg) String() string {
+	return fmt.Sprintf("%v( %v )", oa.ConstFuncOneArgName.String(), oa.Argument.String())
+}
+
+type FEConstFuncOneArgName struct {
```

Comment:
the code gets lengthy with a large number of arg names. Can't we use a string arg name?

Reply:
We want to make sure we capture the right ARG names, otherwise the parser will capture them as fields or values. Also does the job of checking.